extends VisionScpPuppetScript
## SCP-173 puppet script
## Created by Yni, licensed under dual license: for SCP content - GPL 3, for non-SCP - MIT License
class_name Scp173PuppetScript

var rng: RandomNumberGenerator = RandomNumberGenerator.new()
@export var scp_173_variations: Dictionary[PackedScene, int] = {}
var scp_173_current_id: int
@export var invincibility: bool = false
@export var blink_timer_default: float = 4.7
var blink_timer: float = blink_timer_default
var is_blinking: bool = false
var current_human: Node3D
var raycast: RayCast3D
var player_direction: Vector3
var movement_reset: bool = false

# Called when the node enters the scene tree for the first time.
func on_start() -> void:
	raycast = get_parent().get_parent().get_node("RayCastLow")
	#get_parent().get_node("ActionArea").connect("body_entered", on_action_area_body_entered)
	#get_parent().get_node("ActionArea").connect("body_exited", on_action_area_body_exited)
	spawn_scp_variation()
	set_face()


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _physics_process(delta: float) -> void:
	scp_173_blink(delta)
	# If is watching, set velocity to zero, else - go to player.
	if ((is_blinking && watching_puppets.size() > 0 && current_human != null) || (watching_puppets.size() == 0 && current_human != null)) && !freeze:
		scp_173_movement()
		if raycast.is_colliding():
			var collider = raycast.get_collider()
			if collider is MovableNpc:
				if collider.fraction == 0:
					get_parent().get_parent().get_node("InteractSound").stream = load("res://Sounds/Character/Scp173/NeckSnap.ogg")
					get_parent().get_parent().get_node("InteractSound").play()
					collider.health_manage(-16777216)
					active_puppets.erase(current_human)
					current_human = null
					movement_reset = false
## Blink mechanic
func scp_173_blink(delta: float):
	# If blink timer > 0 - then wait
	if blink_timer > 0:
		blink_timer -= delta
	elif !freeze:
		is_blinking = true
		movement_reset = false
		# Navigate to the human near you
		if active_puppets.size() > 0 && !active_puppets.has(current_human):
			current_human = active_puppets[rng.randi_range(0, active_puppets.size() - 1)]
		else:
			current_human = null
		await get_tree().create_timer(0.3).timeout
		blink_timer = blink_timer_default
		is_blinking = false
## Movement control
func scp_173_movement():
	if state == States.IDLE && !movement_reset:
		get_parent().get_parent().set_target_position(current_human.global_position)# + current_human.global_transform.basis.z * 1.5)
		movement_reset = true

func spawn_scp_variation() -> void:
	if scp_173_variations.is_empty():
		get_parent().get_parent().health_manage(-16777216)
	scp_173_current_id = rng.randi_range(0, scp_173_variations.size() - 1)
	var scp_173: Node3D = scp_173_variations.keys()[scp_173_current_id].instantiate()
	add_child(scp_173, true)

## Set face on spawn
func set_face():
	match scp_173_current_id:
		0:
			var tex: ShaderMaterial = load("res://Assets/Materials/Scp173Unity.tres")
			match Settings.current_season:
				Settings.Season.CHRISTMAS:
					tex.set_shader_parameter("albedo_b", load("res://Assets/ExternalModels/SCP/scp173/Faces/face_F1.png"))
				Settings.Season.HALLOWEEN:
					tex.set_shader_parameter("albedo_b", load("res://Assets/ExternalModels/SCP/scp173/Faces/face_H1.png"))
				_:
					tex.set_shader_parameter("albedo_b", load("res://Assets/ExternalModels/SCP/scp173/Faces/face_" + str(rng.randi_range(1, scp_173_variations.values()[scp_173_current_id])) + ".png"))
			get_node("Scp173Entity/SCP173_Rig/Skeleton3D/scp173_MESH").set_surface_override_material(0, tex)
		1:
			var tex: ShaderMaterial = load("res://Assets/Materials/Scp173Our.tres")
			match Settings.current_season:
				Settings.Season.CHRISTMAS:
					tex.set_shader_parameter("albedo_b", load("res://Assets/OriginalModels/Scp173/Faces/FaceF1.png"))
				Settings.Season.HALLOWEEN:
					tex.set_shader_parameter("albedo_b", load("res://Assets/OriginalModels/Scp173/Faces/FaceH1.png"))
				_:
					tex.set_shader_parameter("albedo_b", load("res://Assets/OriginalModels/Scp173/Faces/Face" + str(rng.randi_range(1, scp_173_variations.values()[scp_173_current_id])) + ".png"))
			get_node("Scp173Entity/Cube").set_surface_override_material(0, tex)

#func on_action_area_body_exited(body: Node3D):
	#pass
